### ISSUE TITLE
Building takes very long (slow contribution workflow)

### ISSUE THREAD

**Issue by hasufell** [nylira/prism-break#1633]
Contributing has become slightly annoying, because of 2 things:

1. checking/manipulating all language files
2. building all of them takes a long time, even on a very powerful machine

I've only done haskell web development, so my suggestion would be [hakyll](https://jaspervdj.be/hakyll/), which can use similar human-readable formats for the actual text.

**Comment by strugee** [comment]
1. will be solved by #1627. As for 2... not sure. I might bust out a profiler if I have time and try to speed things up.

**Comment by strugee** [comment]
#1639 landed which allows parallel jobs. Do we still need to keep this open?

**Comment by alerque** [comment]
The build still takes a long time and some tooling for simple repetitive tasks across across languages would be nice, but the build is a lot faster than it used to be (and includes some checks) and baring a rewrite from scratch I don't see it changing dramatically soon. I've been playing around with using `jq` to automate common tasks (see my recent commit messages) but don't have anything setup to share yet.

In any case I'm not sure this issue is serving a purpose any more. If somebody wants to propose a rewrite that's fine, and more specific issues for any commands that would help contributors would be welcome, but this issue is kind of mushy with no real action to take.

**Comment by strugee** [comment]
Yeah, I agree. I wonder if it would help to put a note in either the Pull Request template or `CONTRIBUTING.md` or something like that to remind people that if they run into something specific that's difficult while contributing, they should tell us?

**Comment by hasufell** [comment]
> to remind people that if they run into something specific that's difficult while contributing, they should tell us?

I just did. So why should this issue be closed?

**Comment by strugee** [comment]
> I just did.

Where, in this issue? I don't think there's anything that's specific enough to actually be actionable.

**Comment by hasufell** [comment]
> Where, in this issue? I don't think there's anything that's specific enough to actually be actionable.

First, long build time is **very** specific. Second... just because you don't know how to fix an issue, doesn't make it invalid.

**Comment by alerque** [comment]
> First, long build time is very specific.

"Long" and "very long" are rather subjective terms.  But if I check out this project from the date you opened the issue a full build (not included downloading NPM modules and such) clocks in at 17:02. The current HEAD clocks in at 5:21 on the same machine.

Rebuilding one language from scratch hasn't changed significantly (~0:40) but the considering how the amount of work being done per language has gone up that's still some progress. More importantly there is no longer a hard requirement on rebuilding everything so when you are working on contributions you won't be waiting for rebuilds of things that don't need rebuilding.

Yes it's still a long time. In the end ~80k lines of HTML is generated from ~6k lines of JSON data for each language, and it seems like this could be done faster. If nothing else breaking the build up by output page/category/etc. instead of doing one long series run would probably bring this down significantly without migrating to completely different build system. There might be other obvious things that can be done to cut down on clock cycles too.

My point stands: if somebody wants to overhaul the template pipeline with faster tooling then have at it, but this issue isn't really specific enough. An actionable issue might be "break up input JSON per project/category so build system can only rebuild changed parts" or "generate each output page for a language in parallel threads", or "rewrite template system using Haskell". With action points like that we could discuss and decided if that's the direction we wanted and somebody could actually implement it.

> checking/manipulating all language files

This isn't specific enough either. What checks and manipulations exactly need automating? I've been working on some scripts for my own workflow but I suspect we all contribute through fairly different processes. If specific tasks are known to be hard to  do or take too much time, we can fix them, but I don't know what tasks you're thinking of.

**Comment by hasufell** [comment]
Again: An issue is an issue regardless of whether you know how to solve it. Your definition of "actionable" is pretty uncommon in open source. We don't close issues just because we don't know (yet) how to solve them. Either an issue is *valid* or it is *invalid* or it is *wontfix*. There is no notion of "actionable".

Long build time is a very specific problem that impacts contribution workflow directly and is not theoretical. I don't have to do a performance analysis and show the exact bottleneck of the build process for this to be a specific issue.

> This isn't specific enough either.

It totally is. Look at the PRs. It's a horrible mess and has stopped me numerous times from contributing, because I am wasting time on translation crap.

---

You're saying contributors should tell whether there are problems with the contribution workflow, but as soon as they do you jump to the "details view" and expect specific solutions. That's quite inconsistent.

In big open source projects it is very common to discuss problems in a more high-level manner. It's not always about the details. No one cares what technology prism-break uses. They want to contribute and not be bothered too much. This is a high-level contribution workflow issue, don't confuse high-level with unspecific. Or are you saying you don't care about high-level issues?

That is why there is no reason to close this issue. It can be used to propose, discuss and improve said issues. If you don't care about those issues, then say it and don't pretend you do.

**Comment by alerque** [comment]
> [...] or it is wontfix. There is no notion of "actionable".

I wasn't trying to say build time isn't an issue. Considering how much time I've spent on the build system I'm well aware that this makes a big difference to the viability of the project. All I'm saying is that if we don't plan on doing a re-write then this is `wontfix`. Sorry if my diction didn't fit your pattern.

>  I am wasting time on translation crap.

What translation crap? I believe you. And contributing code that makes it easier for others  to contribute content is something I'm willing to do. But I can't sit down and start writing code based on this description. What task(s) in specific are arduous and what is your current workflow for them?

**Comment by hasufell** [comment]
> But I can't sit down and start writing code based on this description.

I can. As I proposed in this issue description.

Other proposals are welcome too.
